### DISCUSSION TITLE
CI \ CD MAUI - Targeting different environments\release configuration settings

### DISCUSSION THREAD

**Discussion by bozau** [dotnet/maui#11134] [Q&A] [answered]
Hi All,

Just been searching through and not found anything specific to this question at the moment.

I'm trying to setup my CI\CD pipleine (in my case Azure DevOps but it doesn't particularly matter the platform) how would I go about setting up both the MAUI project and build/release pipeline in order to have different configurations with the deployment (CD) targeting different environments e.g. dev, test, or alpha, beta.

Keeping the goal of build once, deploy many times any pointers/tutorials?

Cheers.

Boz.

**Comment by Dreamescaper** [comment, marked as answer]
I'm not sure you'll be able to "build once, deploy many times. Unlike in back-end, you can't change the behavior based on same environment variable, or a separate config file. And your prod APK file probably shouldn't contain any DEV environment URLs or credentials (even if they are not used). So you'll have to create a separate APK file per environment.

As to how to separate configurations per env, I add the following properties to .csproj file:
```xml
    <Environment>Test</Environment>
    <DefineConstants>ENV_$(Environment)</DefineConstants>
```

And create some class with configuration values separated with precompiler directives:
```csharp
    public static class AppSettings
    {
        public static Uri ApiUrl { get; private set; }

        static AppSettings()
        {
#if ENV_Dev
            ApiUrl = new Uri("https://mydev.com");
#elif ENV_Test
            ApiUrl = new Uri("https://mytest.com");
#elif ENV_Production
            ApiUrl = new Uri("https://myprod.com");
#endif
        }
```

And during the build on CI I pass Environment MSBuild variable with a corresponding env name:
`/p:Environment=Production`

**Reply by bozau** [reply]
Appreciate the response, I'll work through this as an approach. 

Another approach that I'm also considering is:

- Build all versions (Dev,Test,Prod), Yes I'd be building for each environment, but, at least it should be a consistent build (same known code base etc)
- Submit them all as Artifacts for the release pipelines
- Release pipeline's control the deployment steps

I've done the above for controlling a NuGet package, not sure if that's something that's against best practice, but seems to be working so far.

Cheers.

Boz

**Comment by IT-VBFK** [comment]
Hi I have similar issues: Mine are regarding proper versioning.

@bozau How did you set up versioning in CI/CD? Especially how to properly version build artifacts (which all should have unique versions)?
I found out, that (speaking for Windows apps here) the `Version` attribute in the `Package.appxmanifest` have to have only versions like `a.b.c.d` where `0 <= a,b,c <= 65535` and `0 <= d <= 9999`

No pre-release tags like `-alphaXX` or `-betaYY` are allowed.

**Reply by Dreamescaper** [reply]
You can try using the latest prerelease version of [Nerdbank.GitVersioning](https://www.nuget.org/packages/Nerdbank.GitVersioning/3.6.65-alpha). It has a nice [functionality](https://github.com/dotnet/Nerdbank.GitVersioning/pull/848) to set MAUI version properties automatically.
